compiler/rustc_const_eval/src/interpret/machine.rs

```diff
@@ -389,11 +389,11 @@ pub trait Machine<'mir, 'tcx: 'mir>: Sized {
     /// need to mutate.
     #[inline(always)]
     fn before_memory_read(
```

Comment:
It's fairly annoying to do (as you can see in the read case). I documented the differences

Reply:
That asymmetry is pretty unsatisfying though and a potential footgun. :/

I think it'd be better to have a new machine hook `before_alloc_access` that gets passed just the AllocId and gets called on reads and writes. Then you also don't need to duplicate the in-bounds check. If the const-eval machine tracks the "bad" AllocId as a local field rather than via the MemoryKind, you don't even need to do the MemoryKind changes.

Reply:
I'll need to pass in whether it's a read or write, too. As we want to allow writes to the root alloc (in order to actually produce the result)

Reply:
Ah, that's fair. Maybe just `before_alloc_read` then. But either way, a separate hook seems to be easier than trying to make the existing hook also do this?